{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import pysam \n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "#* minimap的PAF格式文件\n",
    "#? PAF文件中的坐标是0-base\n",
    "pafData=pd.read_csv(\n",
    "    \"./HC04_A02-Chr03_changeStand.paf\",\n",
    "    usecols=[0,2,3,4,5,7,8],\n",
    "    header=None,index_col=None,sep=\"\\t\"\n",
    ")\n",
    "pafData.columns=[\n",
    "    'queryChrom','Qstart','Qend','stand','referChrom','Rstart','Rend'\n",
    "]\n",
    "pafData=pafData[['referChrom','Rstart','Rend','queryChrom','Qstart','Qend','stand']]\n",
    "pafData=pafData.sort_values(by=['Rstart','Rend'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "#* syri result\n",
    "#! syri中的坐标是1-base\n",
    "syriOut=pd.read_csv(\"./HC04_A01-Chr03_revQuery_changeStandsyri.out\",header=None,index_col=None,sep=\"\\t\",low_memory=False)\n",
    "syriOut=syriOut.loc[(syriOut[5]!=\"-\")&(syriOut[0]!=\"-\")]\n",
    "syriOut=syriOut.astype(\n",
    "    {\n",
    "        1:int,\n",
    "        2:int,\n",
    "        6:int,\n",
    "        7:int\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "#* 根据染色体长度对坐标进行互补\n",
    "def rever_local(chromosomeLength,site):\n",
    "    return chromosomeLength-site+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "def annotateType(x):\n",
    "    '''替换原有的SVs类型\n",
    "    '''\n",
    "    if re.match('INV',x):\n",
    "        if re.match('INV[0-9]*$',x):\n",
    "            return re.sub('INV','SYN',x)\n",
    "        elif re.match('INVAL',x):\n",
    "            return re.sub('INV','SYN',x)\n",
    "        elif re.match('INVTR',x):\n",
    "            return re.sub('INVTR','TRANS',x)\n",
    "        elif re.match('INVDP',x):\n",
    "            return re.sub('INVDP','DUP',x)\n",
    "    elif re.match('SYN',x):\n",
    "        return re.sub('SYN','INV',x) \n",
    "    elif re.match('TRANS',x):\n",
    "        return re.sub('TRANS','INVTR',x)  \n",
    "    elif re.match('DUP',x):\n",
    "        return re.sub('DUP','INVDP',x)  \n",
    "    else:\n",
    "        return x  \n",
    "baseDict={\n",
    "    'A':'T',\n",
    "    'T':'A',\n",
    "    'G':'C',\n",
    "    'C':'G',\n",
    "}        \n",
    "def getBaseInfor(baseInfo,SVType,matchArray,start,end,genomeObject):\n",
    "        if baseInfo!=\"-\":\n",
    "            baseInfo=genomeObject.fetch(reference=matchArray[5],start=start-1,end=end)\n",
    "        #* 判断是否需要互补该基因型\n",
    "        if  re.match(\"INV\",SVType) and len(baseInfo)==1 and baseInfo!=\"-\":\n",
    "                baseInfo=baseDict.get(baseInfo)\n",
    "        return baseInfo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getLocation(PAF_reversed,ChromosomeReverse,chromosomeLength,AlignmentType,genomeObject,syri_queryStart,syri_queryEnd,paf_queryStart,paf_queryEnd,matchArray):\n",
    "    if PAF_reversed==True and ChromosomeReverse==True:\n",
    "        #* query染色体在翻转后，部分染色体的PAF需要修改链才能跑出结果\n",
    "        #? 首先在染色体翻转的情况下，得到翻转下的坐标；然后再进行反向互补\n",
    "        if AlignmentType==\"parent\" :\n",
    "            site1=rever_local(chromosomeLength,syri_queryStart)\n",
    "            site2=rever_local(chromosomeLength,syri_queryEnd)\n",
    "            start=min(site1,site2)\n",
    "            end=max(site1,site2)\n",
    "            #! SVs的类型不需要进行修改\n",
    "            return (\n",
    "               matchArray[0],matchArray[1],\n",
    "                matchArray[2],matchArray[3],\n",
    "                matchArray[4],matchArray[5],\n",
    "                start,end,\n",
    "                matchArray[8],matchArray[9],\n",
    "                matchArray[10],matchArray[11] \n",
    "            )\n",
    "        elif AlignmentType==\"Alignment\":\n",
    "            site1=rever_local(chromosomeLength,syri_queryStart)\n",
    "            site2=rever_local(chromosomeLength,syri_queryEnd)\n",
    "            start=min(site1,site2)\n",
    "            end=max(site1,site2) \n",
    "            return (\n",
    "               matchArray[0],matchArray[1],\n",
    "                matchArray[2],matchArray[3],\n",
    "                matchArray[4],matchArray[5],\n",
    "                start,end,\n",
    "                matchArray[8],matchArray[9],\n",
    "                matchArray[10],matchArray[11] \n",
    "            )\n",
    "        else:\n",
    "            #* 首先需要对坐标进行翻转，再翻转后在进行链的互补\n",
    "            invertStart=-1*(syri_queryStart-(paf_queryStart+1))+paf_queryEnd\n",
    "            invertEnd=-1*(syri_queryEnd-(paf_queryStart+1))+paf_queryEnd\n",
    "            site1=rever_local(chromosomeLength,invertStart)\n",
    "            site2=rever_local(chromosomeLength,invertEnd)\n",
    "            start=min(site1,site2)\n",
    "            end=max(site1,site2)\n",
    "            #* 根据SVs的类型判断第4类的碱基为SNP时是否需要翻转\n",
    "            SVsType=matchArray[9]\n",
    "            baseInfo=matchArray[4]\n",
    "            baseInfo=getBaseInfor(baseInfo,SVsType,matchArray,start,end,genomeObject)\n",
    "            return (\n",
    "                matchArray[0],matchArray[1],\n",
    "                matchArray[2],matchArray[3],\n",
    "                baseInfo,matchArray[5],\n",
    "                start,end,\n",
    "                matchArray[8],matchArray[9],\n",
    "                matchArray[10],matchArray[11] \n",
    "            )\n",
    "    elif PAF_reversed==False and ChromosomeReverse==True:\n",
    "        #* query染色体在翻转后，染色体的PAF不需要修改链,\n",
    "        if AlignmentType==\"parent\" :\n",
    "            site1=rever_local(chromosomeLength,syri_queryStart)\n",
    "            site2=rever_local(chromosomeLength,syri_queryEnd)\n",
    "            start=min(site1,site2)\n",
    "            end=max(site1,site2)\n",
    "            return (\n",
    "               matchArray[0],matchArray[1],\n",
    "                matchArray[2],matchArray[3],\n",
    "                matchArray[4],matchArray[5],\n",
    "                start,end,\n",
    "                annotateType(matchArray[8]),annotateType(matchArray[9]),\n",
    "                annotateType(matchArray[10]),matchArray[11],\n",
    "            )\n",
    "\n",
    "        elif AlignmentType==\"Alignment\":\n",
    "            site1=rever_local(chromosomeLength,syri_queryStart)\n",
    "            site2=rever_local(chromosomeLength,syri_queryEnd)\n",
    "            start=min(site1,site2)\n",
    "            end=max(site1,site2)\n",
    "            return (\n",
    "               matchArray[0],matchArray[1],\n",
    "                matchArray[2],matchArray[3],\n",
    "                matchArray[4],matchArray[5],\n",
    "                start,end,\n",
    "                annotateType(matchArray[8]),annotateType(matchArray[9]),\n",
    "                annotateType(matchArray[10]),matchArray[11],\n",
    "            )\n",
    "        else:\n",
    "            #* 首先需要对坐标进行翻转，再翻转后在进行链的互补\n",
    "            site1=rever_local(chromosomeLength,syri_queryStart)\n",
    "            site2=rever_local(chromosomeLength,syri_queryEnd)\n",
    "            start=min(site1,site2)\n",
    "            end=max(site1,site2)\n",
    "            #* 矫正后的SVs类型\n",
    "            SVsType=annotateType(matchArray[9])\n",
    "            baseInfo=matchArray[4]\n",
    "            baseInfo=getBaseInfor(baseInfo,SVsType,matchArray,start,end,genomeObject)\n",
    "            return (\n",
    "               matchArray[0],matchArray[1],\n",
    "                matchArray[2],matchArray[3],\n",
    "                baseInfo,matchArray[5],\n",
    "                start,end,\n",
    "                annotateType(matchArray[8]),annotateType(matchArray[9]),\n",
    "                annotateType(matchArray[10]),matchArray[11],\n",
    "            )\n",
    "    elif PAF_reversed==True and ChromosomeReverse==False:\n",
    "        #* query 染色体不经过翻转，但是其PAF需要修改链才能继续跑出结果\n",
    "        if AlignmentType==\"parent\" :\n",
    "            start=syri_queryStart\n",
    "            end=syri_queryEnd \n",
    "            return (\n",
    "               matchArray[0],matchArray[1],\n",
    "                matchArray[2],matchArray[3],\n",
    "                matchArray[4],matchArray[5],\n",
    "                start,end,\n",
    "                annotateType(matchArray[8]),annotateType(matchArray[9]),\n",
    "                annotateType(matchArray[10]),matchArray[11],\n",
    "            )\n",
    "        elif AlignmentType==\"Alignment\":\n",
    "            start=syri_queryStart\n",
    "            end=syri_queryEnd\n",
    "            return (\n",
    "               matchArray[0],matchArray[1],\n",
    "                matchArray[2],matchArray[3],\n",
    "                matchArray[4],matchArray[5],\n",
    "                start,end,\n",
    "                annotateType(matchArray[8]),annotateType(matchArray[9]),\n",
    "                annotateType(matchArray[10]),matchArray[11],\n",
    "            )\n",
    "        else:\n",
    "            #* 首先需要对坐标进行翻转\n",
    "            invertStart=-1*(syri_queryStart-(paf_queryStart+1))+paf_queryEnd\n",
    "            invertEnd=-1*(syri_queryEnd-(paf_queryStart+1))+paf_queryEnd\n",
    "            start=min(invertStart,invertEnd)\n",
    "            end=max(invertStart,invertEnd)\n",
    "            #* 矫正后的SVs类型\n",
    "            SVsType=annotateType(matchArray[9])\n",
    "            baseInfo=matchArray[4]\n",
    "            baseInfo=getBaseInfor(baseInfo,SVsType,matchArray,start,end,genomeObject)\n",
    "            return (\n",
    "               matchArray[0],matchArray[1],\n",
    "                matchArray[2],matchArray[3],\n",
    "                baseInfo,matchArray[5],\n",
    "                start,end,\n",
    "                annotateType(matchArray[8]),annotateType(matchArray[9]),\n",
    "                annotateType(matchArray[10]),matchArray[11],\n",
    "            )\n",
    "    else:\n",
    "        #* 正常的情况, 不用使用该脚本处理\n",
    "        pass "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "genomeObject=pysam.FastaFile(\"/cotton/Liuzhenping/Pan-genome/HC04/HC04_V2/HC04_chr_adjust.fa\")\n",
    "out=[]\n",
    "#----------------------------------------------\n",
    "#! 染色体比对时是否翻转了\n",
    "ChromosomeReverse=True\n",
    "#! PAF的链情况是否翻转了\n",
    "PAF_reversed=True\n",
    "chromosomeLength=119507322\n",
    "#----------------------------------------------\n",
    "\n",
    "#* syri的结果是按照顺序排列的\n",
    "paf_queryStart=0\n",
    "paf_queryEnd=0\n",
    "for matchArray in syriOut.values:\n",
    "    if matchArray[3]=='-' and matchArray[4]==\"-\" and matchArray[9]!=\"-\":\n",
    "        #* 该列为Alignment不需要修改对应的坐标\n",
    "        paf_queryStart=min(int(matchArray[6]),int(matchArray[7]))-1\n",
    "        paf_queryEnd=max(int(matchArray[6]),int(matchArray[7])) \n",
    "        newSyriStart=min(int(matchArray[6]),int(matchArray[7]))\n",
    "        newSyriEnd=max(int(matchArray[6]),int(matchArray[7]))\n",
    "        result=getLocation(\n",
    "            PAF_reversed=PAF_reversed,\n",
    "            ChromosomeReverse=ChromosomeReverse,\n",
    "            chromosomeLength=chromosomeLength,\n",
    "            AlignmentType='Alignment',\n",
    "            genomeObject=genomeObject,\n",
    "            syri_queryStart=newSyriStart,\n",
    "            syri_queryEnd=newSyriEnd,\n",
    "            paf_queryStart=paf_queryStart,\n",
    "            paf_queryEnd=paf_queryEnd,\n",
    "            matchArray=matchArray\n",
    "        )\n",
    "    elif matchArray[3]=='-' and matchArray[4]==\"-\" and matchArray[9]==\"-\":\n",
    "        #*该列为parent信息列\n",
    "        newSyriStart=min(int(matchArray[6]),int(matchArray[7]))\n",
    "        newSyriEnd=max(int(matchArray[6]),int(matchArray[7])) \n",
    "        result=getLocation(\n",
    "            PAF_reversed=PAF_reversed,\n",
    "            ChromosomeReverse=ChromosomeReverse,\n",
    "            chromosomeLength=chromosomeLength,\n",
    "            AlignmentType='parent',\n",
    "            genomeObject=genomeObject,\n",
    "            syri_queryStart=newSyriStart,\n",
    "            syri_queryEnd=newSyriEnd,\n",
    "            paf_queryStart=paf_queryStart,\n",
    "            paf_queryEnd=paf_queryEnd,\n",
    "            matchArray=matchArray\n",
    "        )\n",
    "    else:\n",
    "        #* 大区块内的小片段\n",
    "        syri_queryStart=int(matchArray[6])\n",
    "        syri_queryEnd=int(matchArray[7])\n",
    "        result=getLocation(\n",
    "            PAF_reversed=PAF_reversed,\n",
    "            ChromosomeReverse=ChromosomeReverse,\n",
    "            chromosomeLength=chromosomeLength,\n",
    "            AlignmentType='item',\n",
    "            genomeObject=genomeObject,\n",
    "            syri_queryStart=syri_queryStart,\n",
    "            syri_queryEnd=syri_queryEnd,\n",
    "            paf_queryStart=paf_queryStart,\n",
    "            paf_queryEnd=paf_queryEnd,\n",
    "            matchArray=matchArray\n",
    "        )\n",
    "    #todo 修改对应的SVs类型：\n",
    "    out.append(\n",
    "            result\n",
    "        )\n",
    "out=pd.DataFrame(out)\n",
    "\n",
    "\n",
    "#* 存在部分重叠的情况，导致SNP有问题；过滤掉\n",
    "out=out.loc[~((out[3]==out[4])&(out[3]!=\"-\"))]\n",
    "out.to_csv(\"./testsyri_v3.out\",header=False,index=False,sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "out=out.loc[~((out[3]==out[4])&(out[3]!=\"-\"))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr03</td>\n",
       "      <td>38026</td>\n",
       "      <td>49075</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>HC04_A02</td>\n",
       "      <td>108426156</td>\n",
       "      <td>108437331</td>\n",
       "      <td>INVTR5</td>\n",
       "      <td>-</td>\n",
       "      <td>INVTR</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr03</td>\n",
       "      <td>38026</td>\n",
       "      <td>49075</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>HC04_A02</td>\n",
       "      <td>108426156</td>\n",
       "      <td>108437331</td>\n",
       "      <td>INVTRAL18</td>\n",
       "      <td>INVTR5</td>\n",
       "      <td>INVTRAL</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr03</td>\n",
       "      <td>38068</td>\n",
       "      <td>38068</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>HC04_A02</td>\n",
       "      <td>108437289</td>\n",
       "      <td>108437289</td>\n",
       "      <td>SNP17</td>\n",
       "      <td>INVTR5</td>\n",
       "      <td>SNP</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Chr03</td>\n",
       "      <td>38074</td>\n",
       "      <td>38074</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>HC04_A02</td>\n",
       "      <td>108437283</td>\n",
       "      <td>108437283</td>\n",
       "      <td>SNP18</td>\n",
       "      <td>INVTR5</td>\n",
       "      <td>SNP</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Chr03</td>\n",
       "      <td>38088</td>\n",
       "      <td>38088</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>HC04_A02</td>\n",
       "      <td>108437269</td>\n",
       "      <td>108437269</td>\n",
       "      <td>SNP19</td>\n",
       "      <td>INVTR5</td>\n",
       "      <td>SNP</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89257</th>\n",
       "      <td>Chr03</td>\n",
       "      <td>146597309</td>\n",
       "      <td>146597309</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>HC04_A02</td>\n",
       "      <td>102433327</td>\n",
       "      <td>102433327</td>\n",
       "      <td>SNP88424</td>\n",
       "      <td>INV3</td>\n",
       "      <td>SNP</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89258</th>\n",
       "      <td>Chr03</td>\n",
       "      <td>146597314</td>\n",
       "      <td>146597316</td>\n",
       "      <td>AAT</td>\n",
       "      <td>A</td>\n",
       "      <td>HC04_A02</td>\n",
       "      <td>102433322</td>\n",
       "      <td>102433322</td>\n",
       "      <td>DEL88425</td>\n",
       "      <td>INV3</td>\n",
       "      <td>DEL</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89259</th>\n",
       "      <td>Chr03</td>\n",
       "      <td>146597434</td>\n",
       "      <td>146597438</td>\n",
       "      <td>ATATG</td>\n",
       "      <td>A</td>\n",
       "      <td>HC04_A02</td>\n",
       "      <td>102433204</td>\n",
       "      <td>102433204</td>\n",
       "      <td>DEL88426</td>\n",
       "      <td>INV3</td>\n",
       "      <td>DEL</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89260</th>\n",
       "      <td>Chr03</td>\n",
       "      <td>146597440</td>\n",
       "      <td>146597440</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>HC04_A02</td>\n",
       "      <td>102433202</td>\n",
       "      <td>102433202</td>\n",
       "      <td>SNP88427</td>\n",
       "      <td>INV3</td>\n",
       "      <td>SNP</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89261</th>\n",
       "      <td>Chr03</td>\n",
       "      <td>146597450</td>\n",
       "      <td>146597450</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>HC04_A02</td>\n",
       "      <td>102433192</td>\n",
       "      <td>102433192</td>\n",
       "      <td>SNP88428</td>\n",
       "      <td>INV3</td>\n",
       "      <td>SNP</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>89251 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0          1          2      3  4         5          6          7   \\\n",
       "0      Chr03      38026      49075      -  -  HC04_A02  108426156  108437331   \n",
       "1      Chr03      38026      49075      -  -  HC04_A02  108426156  108437331   \n",
       "2      Chr03      38068      38068      G  A  HC04_A02  108437289  108437289   \n",
       "3      Chr03      38074      38074      A  G  HC04_A02  108437283  108437283   \n",
       "4      Chr03      38088      38088      T  C  HC04_A02  108437269  108437269   \n",
       "...      ...        ...        ...    ... ..       ...        ...        ...   \n",
       "89257  Chr03  146597309  146597309      A  G  HC04_A02  102433327  102433327   \n",
       "89258  Chr03  146597314  146597316    AAT  A  HC04_A02  102433322  102433322   \n",
       "89259  Chr03  146597434  146597438  ATATG  A  HC04_A02  102433204  102433204   \n",
       "89260  Chr03  146597440  146597440      A  G  HC04_A02  102433202  102433202   \n",
       "89261  Chr03  146597450  146597450      A  G  HC04_A02  102433192  102433192   \n",
       "\n",
       "              8       9        10 11  \n",
       "0         INVTR5       -    INVTR  -  \n",
       "1      INVTRAL18  INVTR5  INVTRAL  -  \n",
       "2          SNP17  INVTR5      SNP  -  \n",
       "3          SNP18  INVTR5      SNP  -  \n",
       "4          SNP19  INVTR5      SNP  -  \n",
       "...          ...     ...      ... ..  \n",
       "89257   SNP88424    INV3      SNP  -  \n",
       "89258   DEL88425    INV3      DEL  -  \n",
       "89259   DEL88426    INV3      DEL  -  \n",
       "89260   SNP88427    INV3      SNP  -  \n",
       "89261   SNP88428    INV3      SNP  -  \n",
       "\n",
       "[89251 rows x 12 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'AAAAGGGTAGTCCTATACATGCCATTTCAAAGTTCAACCAAAATTGTACCAAAAGGGGGCTTTGATAGTGTGGGAGACTTCGACTTCCAAAAATCCCGAGTCCGATAGCTGACGAGCCAAAATCTATAAAACAGAGAAACAAAGAAACGGAGTAAGCAATTTATGCTTAGTAAGTTTTGAGCAAGAGATTCCAGCACAACAAAAGTATAGCATTCATATAGCTAAACGGATAATTTCATATGCACAAATTTTCAATATCATACTTACTTCACATTACCAACCCTTATGTTCATACACAAAAGATCAACTTAGCCAAAGGCCGGTAGCTCATTTATCGACTGAGCGAATACAGATTTGTAAGGGATCAACTAATTCAAAGCACACACGAAACATACCTCATCGCTGGGATTTTACGAGCGTATTAATTTAAATTTTTACAGCAAGATCGCTCGTTCCCAAATCAAGCATCTTCGGGGTTTAGCCGGATATAACCACTCGCACAAGGCCTTCGGGTCTTAACCCGGATATGGTCACTAGCATAAATGCCTTTGGGACTTAGCCCGGATATAGTCGCTAGCACAAATGCCTTCGGATCTTAGTCCGGATGTAGTCGCTTAGCACAAAAGCCTTCGGGACTTAGCCCGGATATCATTCGAGTAACCATGCACATATATCAATAAATCATGACACATTCATATTTCATTTTCATTACCAAAGCTCAAACACAAGACACTTATCACATTTACAATTTCGGCTCAATAGCCACATATAAGAGCATGATTTTGATTTACTTAAGACATAATCTAATCGAATCATAATTTAAGTCCCATTACTCAAAAACTTACCTCGGATGTTGTCGAACGATTCCGATGGCTATTCGACCACTTTTTCTTTCCCTTTATCGGATTTAGTTCCCCTTTACTCTTGAGCTTAATTTAACAAATAAATTGATTTAATCATTTGAGCATCGAAAAGAGGAACTCAAGGTACTTGGCCCACATATATTCATTAGACATTAAAGTCACATATGTACGGAATCATGAATCAAACTCAACACATTAGTTAATACTCTCCTTAGCCGAATTTTCTAAGTCAAGATAAAGCCTTCAATATGCTTACCTATGGCCGAACAACACATCAACCTATGTACTCATTCATGTGGCCGATTATGCATGTCTATGTTGAGGCCAATTATATACTCAATACCACACACAAATGGCATACCTTTTACTAACTAATGCATTACATATTATAACTCAATACGCATCCATCATGTACTCCATAACCGAAACACCATCATATGTAAATATATACCTTGAGATATTGTATATGTCATACCAATACGTCATGTACAAACATATATATATATGTGCAAGAGCCGAATCTCAAAGTTGATTATAACTAAACATGAACATATATCCAAAACACAAATCTTACCTACCATGCAATAAGCATAAATCATACTTATGGATATACCATGGCCGAATACATCACAACACCATACCATTTCAATTTTGGTCATGGTTAAACAAAGAACTTAATGTCTCACTCAAAAATGCTAAAAAGAAAATCCAAGAGTCATCAATCCACCATCACATGTATCATTAACAAGCTTCATATTTAGCATGCAATGACATTAACACAAAATCCACCTAGGCCGAATATCATCCCCATGACATAGCAAAGATTTGAACCATGGGCTAATTAGAACTCAAGCTAGCAACTAAAAACATACATGAATCTCATGGCACAACCTCAAACATACCTTAATCTAGATACAAGTATGGCCAAACCTCCTCTTAATCCTCTTCCAAACCAAACATGAAGCAAGAACTCCTTCCTCCTTCCTTAGAATTTTCGGCCAAAAGAAATGAAAAAGGATGAACAAATTTTTTTTTCTTTTCTTTAACTCACGGCAATGGGGGGGGGGAACACTCACACACATTCTTTTTTTTTTTTCCATTTCTTTATTACCCATACTCCTTATTTTATTTTTCCTAACATACCTCACTAACATAACATGTTTGTGACATGTTTCCACTCATAGCATGGCCGGCCACTACTAATTAGGGGGAAATTTGACATGCAAGTCCTCCCTTTTTATTACATGCACTATTAGATCCTTATAGATTAGCCTATCACATTTCAAAAGTGTCACACAAATCCTACTAACTGAATTCACATGCAATCGACTAAATCGAAGCTTGAAATTTTCACACATTCATAATTACATATTCTAGACAATAAATATCACATTCAAACATTTCAGTGACTCGGTTTAGCGGTCCCGAAACCACTTCCCGACTAGGGTCAATTTTGGGCTGTCACA'"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genomeObject.fetch(reference=\"HC04_A01\",start=99009224-1,end=99011552)"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "66aa2addb38c89a0d43f63cbdb966cb2dba790ca2bb4fa0a3d79065325ea50af"
  },
  "kernelspec": {
   "display_name": "Python 3.9.13 ('base')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
